src/software/backend/input/network/network_client.h

```diff
@@ -0,0 +1,165 @@
+     * @param received_vision_detection_callback This function will be called with a new
+     * VisionDetection every time one is received
+     * @param received_refbox_data_callback This function will be called with a new
+     * RefboxData every time one is received
+     */
+    explicit NetworkClient(
+        std::string vision_multicast_address, int vision_multicast_port,
+        std::string gamecontroller_multicast_address, int gamecontroller_multicast_port,
+        std::function<void(VisionDetection)> received_vision_detection_callback,
+        std::function<void(RefboxData)> received_refbox_data_callback);
+
+    /**
+     * Safely destructs this NetworkClient object. Stops any running IO services and
+     * gracefully joins the thread before exiting.
+     */
+    ~NetworkClient();
+
+    // Delete the copy and assignment operators because this class really shouldn't need
+    // them and we don't want to risk doing anything nasty with the internal
+    // threading this class uses
+    NetworkClient& operator=(const NetworkClient&) = delete;
+    NetworkClient(const NetworkClient&)            = delete;
+
+   private:
+    /**
+     * Sets up the vision client to the point where it will receive packets on the
+     * given address/port
+     *
+     * @param vision_address String representation of the vision IP address
+     * @param vision_port The port vision is running on
+     */
+    void setupVisionClient(std::string vision_address, int vision_port);
+
+    /**
+     * Sets up the gamecontroller client to the point where it will receive packets on the
+     * given address/port
+     *
+     * @param gamecontroller_address String representation of the gamecontroller IP
+     * address
+     * @param gamecontroller_port The port gamecontroller is running on
+     */
+    void setupGameControllerClient(std::string gamecontroller_address,
+                                   int gamecontroller_port);
+
+    /**
+     * Starts up the IO service thread to run and service network requests in the
+     * background
+     */
+    void startIoServiceThreadInBackground();
+
+    /**
+     * Given a detection, figures out if the camera is enabled
+     *
+     * @param detection SSL_DetectionFrame to consider
+     *
+     * @return whether the camera is enabled
+     */
+    bool isCameraEnabled(const SSL_DetectionFrame& detection);
+
+    // TODO: Remove this wrapper function once we move to a better simulator
+    // https://github.com/UBC-Thunderbots/Software/issues/609
+    /**
+     * A wrapper function for the filterAndPublishVisionData function. This wrapper is
+     * responsible for ignoring any bad packets we get from grSim, because grSim
+     * sends garbage packets from very far in the future that causes issues if they
+     * get through to our filters and logic.
+     *
+     * @param packet The vision packet
+     */
+    void filterAndPublishVisionDataWrapper(SSL_WrapperPacket packet);
```

Comment:
nit: can we rename all these `filterAnd...` functions to remove the `filter` part? (since there's no longer any filtering here)

Reply:
changed